<a href="https://colab.research.google.com/github/LatiefDataVisionary/numpy-with-w3schools/blob/main/numpy_ufuncs.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. ufuncs Intro**

Without ufunc, we can use Python's built-in zip() method:



In [2]:
x = [1,2,3,4]
y = [4,5,6,7]
z=[]

for i, j in zip(x,y):
  z.append(i+j)
z

[5, 7, 9, 11]

With ufunc, we can use the add() function:



In [3]:
import numpy as np

x = [1, 2, 3, 4]
y = [4, 5, 6, 7]
z = np.add(x,y)
z

array([ 5,  7,  9, 11])

# **2. ufuncs Create Function**

Create your own ufunc for addition:



In [4]:
def myadd(x,y):
  return x+y

myadd = np.frompyfunc(myadd, 2, 1)
myadd([1,2,3,4], [5,6,7,8])

array([6, 8, 10, 12], dtype=object)

Check if a function is a ufunc:



In [5]:
type(np.add)

numpy.ufunc

Check the type of another function: concatenate():



In [6]:
type(np.concatenate)

numpy._ArrayFunctionDispatcher

Use an if statement to check if the function is a ufunc or not:



In [7]:
if type(np.add) == np.ufunc:
  print('add is ufunc')
else:
  print('add is not ufunc')

add is ufunc


# **3. Simple Arithmetic**


Add the values in arr1 to the values in arr2:



In [8]:
arr1 = np.array([10, 11, 12, 13, 14, 15])
arr2 = np.array([20, 21, 22, 23, 24, 25])

newarr = np.add(arr1, arr2)
newarr

array([30, 32, 34, 36, 38, 40])

Subtract the values in arr2 from the values in arr1:



In [10]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([20, 21, 22, 23, 24, 25])

newarr = np.subtract(arr1, arr2)
newarr

array([-10,  -1,   8,  17,  26,  35])

Multiply the values in arr1 with the values in arr2:



In [11]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([20, 21, 22, 23, 24, 25])

newarr = np.multiply(arr1, arr2)
newarr

array([ 200,  420,  660,  920, 1200, 1500])

Divide the values in arr1 with the values in arr2:



In [15]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 5, 10, 8, 2, 33])

newarr = np.round(np.divide(arr1, arr2), 2)
newarr

array([ 3.33,  4.  ,  3.  ,  5.  , 25.  ,  1.82])

Raise the valules in arr1 to the power of values in arr2:



In [16]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 5, 6, 8, 2, 33])

newarr = np.power(arr1, arr2)
newarr

array([         1000,       3200000,     729000000, 6553600000000,
                2500,             0])

Both the mod() and the remainder() functions return the remainder of the values in the first array corresponding to the values in the second array, and return the results in a new array.



In [17]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.mod(arr1, arr2)
newarr

array([ 1,  6,  3,  0,  0, 27])

In [18]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.remainder(arr1, arr2)
newarr

array([ 1,  6,  3,  0,  0, 27])

In [19]:
# Contoh dengan bilangan negatif
arr1 = np.array([-10, 10])
arr2 = np.array([3, -3])

print("Modulus (mod):", np.mod(arr1, arr2))
print("Remainder (remainder):", np.remainder(arr1, arr2))

Modulus (mod): [ 2 -2]
Remainder (remainder): [ 2 -2]


Return the quotient and mod:



In [20]:
arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.divmod(arr1, arr2)
newarr

(array([ 3,  2,  3,  5, 25,  1]), array([ 1,  6,  3,  0,  0, 27]))

Return the absolute values:



In [21]:
arr = np.array([-1, -2, 1, 2, 3, -4])
newarr = np.absolute(arr)
newarr

array([1, 2, 1, 2, 3, 4])

In [22]:
arr = np.array([-1, -2, 1, 2, 3, -4])
newarr = np.abs(arr)
newarr

array([1, 2, 1, 2, 3, 4])

# **4. ufunc Rounding Decimals**


Truncate elements of following array:



In [23]:
arr = np.trunc([-3.1666, 3.6667])
arr

array([-3.,  3.])

Same example, using fix():



In [24]:
arr = np.fix([-3.6667, 3.6667])
arr

array([-3.,  3.])

Round off 3.1666 to 2 decimal places:



In [25]:
arr = np.around(3.1666, 2)
arr

3.17

Floor the elements of following array:



In [26]:
arr = np.floor([-3.1666, 3.6667])
arr

array([-4.,  3.])

Ceil the elements of following array:



In [27]:
arr = np.ceil([-3.1666, 3.6667])
arr

array([-3.,  4.])

# **5. ufunc NumPy Logs**


Find log at base 2 of all elements of following array:



In [33]:
arr = np.arange(1, 10)
print(arr)
print(np.log2(arr))

[1 2 3 4 5 6 7 8 9]
[0.         1.         1.5849625  2.         2.32192809 2.5849625
 2.80735492 3.         3.169925  ]


Find log at base 10 of all elements of following array:



In [36]:
arr = np.arange(1, 11)
print(arr)
print(np.log10(arr))

[ 1  2  3  4  5  6  7  8  9 10]
[0.         0.30103    0.47712125 0.60205999 0.69897    0.77815125
 0.84509804 0.90308999 0.95424251 1.        ]


Find log at base e of all elements of following array:



In [37]:
arr = np.arange(1, 10)
np.log(arr)

array([0.        , 0.69314718, 1.09861229, 1.38629436, 1.60943791,
       1.79175947, 1.94591015, 2.07944154, 2.19722458])

Log at Any Base


In [39]:
from math import log

nplog = np.frompyfunc(log, 2, 1)
nplog(100, 15)

1.7005483074552052

Menghitung logaritma 100 dengan basis 15.
Hasilnya adalah sekitar 1.7005483074552052, karena
15
1.7005
≈
100
15
1.7005
 ≈100

# **6. NumPy Summations**


Sum the values in arr1 and the values in arr2:



In [40]:
arr1 = np.array([1, 2, 3])
arr2 = np.array([1, 2, 3])

newarr = np.sum([arr1, arr2])
newarr

12

Perform summation in the following array over 1st axis:



In [44]:
arr1 = np.array([1, 2, 3])
arr2 = np.array([1, 2, 3])

newarr_ax1 = np.sum([arr1, arr2], axis=1)
print(newarr_ax1)

[6 6]


Perform cummulative summation in the following array:



In [45]:
arr = np.array([1, 2, 3])
newarr = np.cumsum(arr)
newarr

array([1, 3, 6])

Cummulative sum means partially adding the elements in array.

E.g. The partial sum of [1, 2, 3, 4] would be [1, 1+2, 1+2+3, 1+2+3+4] = [1, 3, 6, 10].

Perfom partial sum with the cumsum() function.